[addons][viz] code cleanup / improvements / API update

AlwinEsch opened the pull request:
## Description

This cleans up the visualization add-on system by take of code from my sandbox project. Also for here relates not direct to this project and can be brought in independent.

Main change is to remove the fourier transformation inside Kodi and do only by add-ons (was only one still there, where used them from Kodi).
The other is to cleanup the API and have better in overview.

All related changes are separated in own commits with short descriptions about.

This should be the last API change added from me for Nexus and want to make all others in coming Kodi O*.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
With changed add-ons.
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][viz] remove functions to get add-on name and paths

This can also be done with "kodi::addon::GetUserPath()", "kodi::addon::GetAddonPath()" and "kodi::addon::GetAddonInfo()".
Is to have more clean and reduced in not needed codes.

AlwinEsch committed:
[addons][viz] clang cleanup about visualization code parts

To have on following commits and future more clean.

AlwinEsch committed:
[addons][viz] change "C" function names inside CInstanceVisualization

This done to match his name to "C" style used inside related function table "KodiToAddonFuncTable_Visualization"..

AlwinEsch committed:
[addons][viz] don't use fourier transformed audio in API

The fourier system used is imprecise, which means that it does not appear correct in the converted display, and on the other hand, unnecessary code because most add-ons do not use it (either audio data or FFTW in it itself).

AlwinEsch committed:
[addons][viz] replaced AddonProps_Visualization with callback function

The background here is also a planned sandbox system. Because the properties must be available first, it is difficult to have them defined before the add-on was created. Therefore it is hereby replaced with a callback.
In the C ++ header the previous way remains, once it calls the new "C" function to get the data.

AlwinEsch committed:
[addons][viz] change "std::string" inside header to "const std::string&"

Done to prevent not needed malloc/copy works on the add-on.

AlwinEsch committed:
[addons][viz] Used typedef to define "C" function pointer

This has the background to be able to use several parts in the sandbox with automatic creation.
It also makes the structure look cleaner.

In the commit, the value names are also changed to lowercase, which corresponds more to the "C" style.

AlwinEsch committed:
[addons][viz] increase API to version 4.0.0

Used on ADDON_INSTANCE_VERSION_VISUALIZATION and ADDON_INSTANCE_VERSION_VISUALIZATION_MIN

AlwinEsch committed:
[addons][viz] make callback with internal functions

On the one hand, this has the background to be ready for the sandbox, where a script determines the callback code and, on the other hand, to not use any static functions in the class. Instead, they are now stored in an anonymous namespace (they are removed in the sandbox).

AlwinEsch committed:
[addons][viz] take addon pointer directly to calls

Before that, the complete "C" structure pointer was passed, which is actually not necessary (and also unnecessary work to get pointers).
This changes it to only pass the pointer required by the add-on.

garbear commented on xbmc/guilib/GUIVisualisationControl.h:
```diff
@@ -95,5 +95,5 @@ class CGUIVisualisationControl : public CGUIControl, public IAudioCallback
   std::string m_profilePath; /*!< To add-on sended profile path */
 
-  ADDON::CVisualization* m_instance{nullptr};
+  KODI::ADDONS::CVisualization* m_instance{nullptr};
```
`#include "addons/Visualization.h"` can be moved to the cpp file and this class forward-declared.

AlwinEsch replied:
Thanks for the note, will change it.
Also you brought me to an idea to change it to `std::unique_ptr` to confirm destruct in all cases.

garbear reviewed (approved):
Nice cleanup. I see how the previous PRs progressed to this one.

Only comment would be to clang-format the entire xbmc/addons (you've earned this folder) and be done with it.

AlwinEsch committed:
[addons][viz] change namespace from ADDON to KODI::ADDONS

To be more correct to Kodi's current namespace style.

AlwinEsch committed:
[addons][viz] code cleanup on CGUIVisualisationControl

Done to make a bit more optimized and to remove not needed parts.

AlwinEsch committed:
[addons][viz] use CVisualization as std::unique_ptr

This stores the "KODI::ADDONS::CVisualization" within an std::unique_ptr to confirm in all cases an desctruction about.

fuzzard commented:
Jenkins build this please
